src/cli/field_data.rs

```diff
@@ -0,0 +1,74 @@
+use anyhow::Result;
+
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct FieldData<T>(T);
+
+pub(crate) trait HasFieldModulus {
+    fn field_modulus() -> String;
+}
+
+impl<'de, T: DeserializeOwned + HasFieldModulus> Deserialize<'de> for FieldData<T> {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct Labeled {
+            label: String,
+            #[serde(with = "serde_bytes")]
+            bytes: Vec<u8>,
+        }
+        let level1_struct = Labeled::deserialize(deserializer)?;
+        if level1_struct.label != T::field_modulus() {
+            return Err(serde::de::Error::custom("Field mismatch"));
+        };
+        let t: T =
+            bincode::deserialize(&level1_struct.bytes[..]).map_err(serde::de::Error::custom)?;
+        Ok(FieldData(t))
+    }
+}
+
+impl<T: Serialize + HasFieldModulus> Serialize for FieldData<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        #[derive(Serialize)]
+        struct Labeled {
+            label: String,
+            #[serde(with = "serde_bytes")]
+            bytes: Vec<u8>,
+        }
+        let bytes = bincode::serialize(&self.0).map_err(serde::ser::Error::custom)?;
+        let labeled = Labeled {
+            label: T::field_modulus(),
+            bytes,
+        };
+        labeled.serialize(serializer)
+    }
+}
+
```

Comment:
We might add a couple unit tests here, which could both serve as a round-trip test (though another way is to implement `Arbitrary` for `FieldData` and use the `serde_test` macro), and document the intent of this struct. These tests are the right place to make your valuable point about sequential deserialization steps (e.g. in code comments).

Reply:
Can we leave that for another PR? I can open an issue for it